core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -288,11 +292,10 @@ class ReplicaFetcherThread(name: String,
         requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.offset, fetchSize))
     }
 
-    val request =
-      if (fetchRequestVersion >= 3) JFetchRequest.fromReplica(replicaId, maxWait, minBytes, maxBytes, requestMap)
-      else JFetchRequest.fromReplica(replicaId, maxWait, minBytes, requestMap)
-
-    new FetchRequest(request)
+    val requestBuilder = new JFetchRequest.Builder(maxWait, minBytes, requestMap).
+        setReplicaId(replicaId).setMaxBytes(maxBytes)
+    requestBuilder.setVersion(fetchRequestVersion)
```

Comment:
Oh, I see. The compiler is correct in not allowing that then. One way to fix that is to change the definition of `Builder` to be:

```java
public static abstract class Builder<T extends AbstractRequest, B extends Builder<T, B>> {
  ...
  public B setVersion(short version) {
  ...
```

But this makes it more awkward to use the abstract builder directly due to the additional type parameter (in Scala we could use type members instead). An alternative is to override `setBuilder` in each builder and use the more specific type there. It's a bit annoying, but builders are supposedly used more times than they are defined.